scipy/sparse/linalg/isolve/tests/test_iterative.py

```diff
@@ -207,6 +205,8 @@ def check_convergence(solver, case):
 
     if A.dtype.char in "dD":
         tol = 1e-8
+    elif solver is minres and (A.dtype.char == 'F'):
```

Comment:
So the complex case works with minor changes, but with a tolerance of `1e-2` instead of `1e-8` for double. Is that expected?

Reply:
Thank you for your review.
> So the complex case works with minor changes

Yes, for MINRES, it is just a few small changes expand from real problem to complex problem.

> but with a tolerance of 1e-2 instead of 1e-8 for double. Is that expected?

Yes, In fact, the upper bound of the tolerence for these cases is about 4.0e-2 ~ 3.0e-2 (if A.dtype.char is not in "dD"), so in original code `tol=1e-2`  holds. whereas complex MINRES's upper bound here is about 5.0e-4 ~ 4.0e-4, so `tol=1e-4` is OK, but I set `tol=1e-8` here, it's also OK because of expecting for higher accuracy.

Reply:
Let me add some details. With a lot of iterative algorithm, its convergence criterion usually doesn't directly use `||b - Ax||/||b|| < tol` (or equivalently use `||b - Ax|| < tol * ||b||` in some books, where `tol * ||b||` is so-called `atol`), because it causes compute "1 MatMultVec + 1 AXPY + 1 NORM" each iteration, which are very expensive, so people use a cheap way. The cheap way is not equivalent to true convergence criterion in some algorithm such as TFQMR/GMRES/MINRES algorithms etc. In some cases, if "tol" is large, the true condition "||b-Ax|| < tol * ||b||" doesn't hold, so in  the original code you can find `tol=1e-2` instead of `tol=1e-1`  or larger "tol" which will cause failed.

Reply:
Thanks for the explanation @zhaog6.

Just to spell it out a bit more (these character codes are always hard to read);
- `dD` means: `float64`, `complex128`. These both have 1e-8 tolerance. 
- `F` is `complex64`. This now also has 1e-8.
- `f` or `float32` is missing, so that's still captured by the `else:` clause and has 1e-2.

So what seems odd is that `complex64` has an accuracy much higher than `float32`. Could you add the `float32` case explicitly?

> whereas complex MINRES's upper bound here is about 5.0e-4 ~ 4.0e-4, so `tol=1e-4` is OK, but I set `tol=1e-8` here, it's also OK because of expecting for higher accuracy.

I'm not sure I understand this. Are you saying that `1e-8` just happens to pass for this particular test case, but for different test cases that may not hold?

Reply:
> So what seems odd is that complex64 has an accuracy much higher than float32. Could you add the float32 case explicitly?
I just want to verify it also holds for more accurate "tol" before, so choosing `tol=1e-8` for complex64 MINRES. but `tol=1e-4` is more proper, I will change `tol=1e-8` into `1e-4` (the upper bound for `complex64`) if need to change, that is
```python
if A.dtype.char in "dD":
    tol = 1e-8
elif (solver is minres) and (A.dtype.char =='F'):
    tol = 1e-4
else:
    tol = 1e-2
```
With the `float32` (`f`) case, it has been included in the `else` branch of the code above. Thanks

Reply:
> Are you saying that 1e-8 just happens to pass for this particular test case, but for different test cases that may not hold?

No, for all cases, `tol=1e-8` still holds. I just want to verify the fact that the PR (MINRES for complex systems) is also convergent to more tougher accuracy than "tol=1e-4". Thanks

Reply:
Dear Ralf @rgommers I also want to add some details to help understand. In the original MINRES file, the original implementation used a special convergence criterion, that is: Firstly for convenience, we define "newNorm" as `max{ ||r||/(||A||*||x||), ||Ar||/(||A||*||r||) }` for `Ax = b`, and then the special criterion used in `minres.py` is `newNorm < tol`. But the standard criterion is `||b - Ax||/||b|| < tol` which is the exactly criterion in `tests/test_iterative.py", whereas we know "newNorm“ is not equal to `||b - Ax||/||b||`. So the difference will affects the convergence history of the extreme cases (because use the special norm "newNorm" instead of the standard norm). which is why taking `tol=1e-2` or the smaller instead of `tol=1e-1` or the larger in the `else` branch. But for `complex64` MINRES, we can take `tol=1e-4` or the smaller such as `tol=1e-6`, `1e-8`.

Reply:
Hi. @zhaog6 I am interested in knowing what's the actual error in the outputs of MINRES in your branch? Like, for the current test cases what's the error between the expected output and the output given by algorithm after your changes. Thanks.

Reply:
> Hi. @zhaog6 I am interested in knowing what's the actual error in the outputs of MINRES in your branch? Like, for the current test cases what's the error between the expected output and the output given by algorithm after your changes. Thanks.

The actual error outputs for MINRES in `minres.py` are `||r||/(||A||*||x||)` and `||Ar||/(||A||*||r|)|`, and the convergence criterion is one of `||r||/(||A||*||x||)` and `||Ar||/(||A||*||r||)` is less than `tol`, you can find it in line 302-305 in the button "Files changed", that is
```python
if test2 <= tol:
    istop = 2
if test1 <= tol:
    istop = 1
```
AND
```python
# in line 340-341
if istop != 0:
    break
```
But, in the test file `test_iterative.py`, the error between the expected output and the output is measured by `||b - Ax||/||b|| < tol` instead of `test1 <= tol` or `test2 <= tol`. the convergerce condition in `minres.py` and in `test_iterative.py` is not equivalent.
the PR doesn't change the definition of error, just to add `conjugate` to support complex systems.